File: src/tests/javaobject/newapi/Portfolio.java
/*=========================================================================
 * Copyright (c) 2010-2014 Pivotal Software, Inc. All Rights Reserved.
 * This product is protected by U.S. and international copyright
 * and intellectual property laws. Pivotal products are covered by
 * one or more patents listed at http://www.pivotal.io/patents.
 *=========================================================================
 */
//package org.apache.geode.cache.query.data;
package javaobject.newapi;


import java.util.*;
import java.io.*;
import org.apache.geode.*; // for DataSerializable
import org.apache.geode.cache.Declarable;


public class Portfolio implements Declarable, Serializable, DataSerializable {

  public int ID;
  public String pkid;
  public Position position1;
  public Position position2;
  public HashMap positions = new HashMap();
  public String type;
  public String status;
  public String [] names={"aaa","bbb","ccc","ddd"};
  public byte[] newVal;
  public Date creationDate = new Date();
  public byte[] arrayZeroSize;
  public byte[] arrayNull;

  /**
   * Initializes an instance of <code>Portfolio</code> from a
   * <code>Properties</code> object assembled from data residing in a
   * <code>cache.xml</code> file.
   */
  public void init(Properties props) {
    this.ID = Integer.parseInt(props.getProperty("ID"));
    this.pkid = props.getProperty("pkid");
    this.type = props.getProperty("type", "type1");
    this.status = props.getProperty("status", "active");
    // get the positions. These are stored in the properties object
    // as Positions, not String, so use Hashtable protocol to get at them.
    // the keys are named "positionN", where N is an integer.
    for (Iterator itr = props.entrySet().iterator(); itr.hasNext(); ) {
      Map.Entry entry = (Map.Entry)itr.next();
      String key = (String)entry.getKey();
      if (key.startsWith("position")) {
        Position pos = (Position)entry.getValue();
        this.positions.put(pos.getSecId(), pos);
      }
    }
  }

  public Portfolio select(){
    return this;
  }
  public int getID() {
    return ID;
  }

  public String getPk() {
    return pkid;
  }
  public HashMap getPositions() {
    return positions;
  }

  public HashMap getPositions(String str) {
    return positions;
  }

  public HashMap getPositions(Integer i) {
    return positions;
  }

  public HashMap getPositions(int i) {
    return positions;
  }

  public Position getP1() {
    return position1;
  }

  public Position getP2() {
    return position2;
  }

  public Date getCreationDate() {
    return creationDate;
  }

  public boolean testMethod(boolean booleanArg) {
    return true;
  }

  public boolean isActive() {
    return status.equals("active");
  }
  public byte[] getNewVal(){
    return this.newVal;
  }
  public byte[] ArrayZeroSize(){
      return this.arrayZeroSize;
  }
  public byte[] ArrayNull(){
      return this.arrayNull;
  }

  static {
     Instantiator.register(new Instantiator(Portfolio.class, (byte) 8) {
     public DataSerializable newInstance() {
        return new Portfolio();
     }
   });
 }


  public static String secIds[] = { "SUN", "IBM", "YHOO", "GOOG", "MSFT",
      "AOL", "APPL", "ORCL", "SAP", "DELL", "RHAT", "NOVL", "HP"};

  /* public no-arg constructor required for Deserializable */
  public Portfolio() {
  }

  public Portfolio(int i, int size) {
    ID = i;
    pkid = "" + i;
    status = i % 2 == 0 ? "active" : "inactive";
    type = "type" + (i % 3);
    position1 = new Position(secIds[Position.cnt % secIds.length],
        Position.cnt * 1000);
    if (i % 2 != 0) {
      position2 = new Position(secIds[Position.cnt % secIds.length],
          Position.cnt * 1000);
    }
    else {
      position2 = null;
    }
    positions.put(secIds[Position.cnt % secIds.length], new Position(secIds[Position.cnt % secIds.length], Position.cnt * 1000));
    newVal = new byte[size];
    for(int index = 0; index < size; index++) {
      newVal[index] = 'B';
    }
    arrayZeroSize = new byte[0];
    arrayNull = null;
  }

  public String toString() {
    String out = "Portfolio [ID=" + ID + " status=" + status + " type=" + type
        + "pkid=" + pkid + "creationDate=" + creationDate + "\n ";
    Iterator iter = positions.entrySet().iterator();
    while (iter.hasNext()) {
      Map.Entry entry = (Map.Entry) iter.next();
      out += entry.getKey() + ":" + entry.getValue() + ", ";
    }
    out += "\n P1:" + position1 + ", P2:" + position2;
    return out + "\n]";
  }

  /**
   * Getter for property type.S
   *
   * @return Value of property type.
   */
  public String getType() {
    return this.type;
  }

  public boolean boolFunction(String strArg){
      if(strArg=="active"){
      return true;
      }
      else{
          return false;
      }
  }

  public static void writeHashmap( DataOutput output, HashMap map ) throws IOException
  {
    output.writeInt( map.size() );
    for (Iterator iter = map.entrySet().iterator();iter.hasNext();) {
      Map.Entry entry = (Map.Entry) iter.next();
      String mykey = (String)entry.getKey();
      DataSerializer.writeString(mykey,output);
      DataSerializer.writeObject(entry.getValue(),output);
    }
  }

  public static HashMap readHashmap( DataInput input ) throws IOException, ClassNotFoundException
  {
    HashMap map = new HashMap( );
    int mapSize = input.readInt();
    for (int i = 0; i < mapSize; i++) {
      String key = DataSerializer.readString(input);
      Position pos = (Position)DataSerializer.readObject(input);
      map.put(key,pos);
    }
    return map;
  }

  public void fromData(DataInput in) throws IOException, ClassNotFoundException {
    this.ID = in.readInt();
	  this.pkid = in.readUTF();
    this.position1 = (Position)DataSerializer.readObject(in);
    this.position2 = (Position)DataSerializer.readObject(in);
    this.positions = (HashMap)DataSerializer.readHashMap(in);
    this.type = in.readUTF();
    this.status = in.readUTF();
    this.names = (String[])DataSerializer.readObject(in);
    this.newVal = (byte[])DataSerializer.readByteArray(in);
    this.creationDate = DataSerializer.readDate(in);
    this.arrayNull = DataSerializer.readByteArray(in);
    this.arrayZeroSize = DataSerializer.readByteArray(in);
  }
  
  public void toData(DataOutput out) throws IOException {
    out.writeInt(this.ID);
	  out.writeUTF(this.pkid);
    DataSerializer.writeObject(this.position1, out);
    DataSerializer.writeObject(this.position2, out);
	  DataSerializer.writeHashMap(this.positions,out);
	  out.writeUTF(this.type);
    out.writeUTF(this.status);
    DataSerializer.writeObject(this.names, out);
    DataSerializer.writeByteArray(this.newVal,out);
    DataSerializer.writeDate(this.creationDate,out);
    DataSerializer.writeByteArray(this.arrayNull, out);
    DataSerializer.writeByteArray(this.arrayZeroSize, out);
  }
  
  public static boolean compareForEquals(Object first, Object second) {
    if (first == null && second == null) return true;
    if (first != null && first.equals(second)) return true;
    return false;
  }

  public boolean equals(Object other) {
    if (other==null) return false;
    if (!(other instanceof Portfolio)) return false;

    Portfolio port = (Portfolio) other;

    if (this.ID != port.ID) return false;

    if (!Portfolio.compareForEquals(this.pkid, port.pkid)) return false;
    if (!Portfolio.compareForEquals(this.position1, port.position1)) return false;
    if (!Portfolio.compareForEquals(this.position2, port.position2)) return false;
    if (!Portfolio.compareForEquals(this.positions, port.positions)) return false;
    if (!Portfolio.compareForEquals(this.type, port.type)) return false;
    if (!Portfolio.compareForEquals(this.status, port.status)) return false;
    if (!Portfolio.compareForEquals(this.names, port.names)) return false;
    if (!Portfolio.compareForEquals(this.newVal, port.newVal)) return false;
    if (!Portfolio.compareForEquals(this.creationDate, port.creationDate)) return false;

    return true;
  }

  public int hashCode() {
    int hashcode = ID;
    if (this.pkid != null) hashcode ^= pkid.hashCode();
    if (this.position1 != null) hashcode ^= position1.hashCode();
    if (this.position2 != null) hashcode ^= position2.hashCode();
    if (this.positions != null) hashcode ^= positions.hashCode();
    if (this.type != null) hashcode ^= type.hashCode();
    if (this.status != null) hashcode ^= status.hashCode();
    if (this.names != null) hashcode ^= names.hashCode();
    if (this.newVal != null) hashcode ^= newVal.hashCode();
    if (this.creationDate != null) hashcode ^= creationDate.hashCode();
    return hashcode;
  }
}

File: geode-client-native/src/tests/clicache/NewTestObject/Portfolio.cs
//=========================================================================
// Copyright (c) 2002-2014 Pivotal Software, Inc. All Rights Reserved.
// This product is protected by U.S. and international copyright
// and intellectual property laws. Pivotal products are covered by
// more patents listed at http://www.pivotal.io/patents.
//========================================================================

using System;
using System.Collections.Generic;
namespace GemStone.GemFire.Cache.Tests.NewAPI
{

  using GemStone.GemFire.Cache.Generic;
  /// <summary>
  /// User class for testing the put functionality for object.
  /// </summary>
  public class Portfolio
    : IGFSerializable
  {
    #region Private members and methods

    private int m_id;
    private string m_pkid;
    private Position m_position1;
    private Position m_position2;
    private Dictionary<Object, Object> m_positions;
    private string m_type;
    private string m_status;
    private string[] m_names;
    private byte[] m_newVal;
    private DateTime m_creationDate;
    private byte[] m_arrayZeroSize;
    private byte[] m_arrayNull;

    private static string[] m_secIds = { "SUN", "IBM", "YHOO", "GOOG", "MSFT",
      "AOL", "APPL", "ORCL", "SAP", "DELL" };

    private UInt32 GetObjectSize(IGFSerializable obj)
    {
      return (obj == null ? 0 : obj.ObjectSize);
    }

    #endregion

    #region Public accessors

    public int ID
    {
      get
      {
        return m_id;
      }
    }

    public string Pkid
    {
      get
      {
        return m_pkid;
      }
    }

    public Position P1
    {
      get
      {
        return m_position1;
      }
    }

    public Position P2
    {
      get
      {
        return m_position2;
      }
    }

    public IDictionary<Object, Object> Positions
    {
      get
      {
        return m_positions;
      }
    }

    public string Status
    {
      get
      {
        return m_status;
      }
    }

    public bool IsActive
    {
      get
      {
        return (m_status == "active");
      }
    }

    public byte[] NewVal
    {
      get
      {
        return m_newVal;
      }
    }

    public byte[] ArrayNull
    {
      get
      {
        return m_arrayNull;
      }
    }

    public byte[] ArrayZeroSize
    {
      get
      {
        return m_arrayZeroSize;
      }
    }

    public DateTime CreationDate
    {
      get
      {
        return m_creationDate;
      }
    }

    public string Type
    {
      get
      {
        return m_type;
      }
    }

    public static string[] SecIds
    {
      get
      {
        return m_secIds;
      }
    }

    public override string ToString()
    {
      string portStr = string.Format("Portfolio [ID={0} status={1} " +
        "type={2} pkid={3}]", m_id, m_status, m_type, m_pkid);
      portStr += string.Format("{0}\tP1: {1}",
        Environment.NewLine, m_position1);
      portStr += string.Format("{0}\tP2: {1}",
        Environment.NewLine, m_position2);
      portStr += string.Format("{0}Creation Date: {1}", Environment.NewLine,
        m_creationDate);
      return portStr;
    }

    #endregion

    #region Constructors

    public Portfolio()
    {
      m_id = 0;
      m_pkid = null;
      m_type = null;
      m_status = null;
      m_newVal = null;
      m_creationDate = DateTime.MinValue;
    }

    public Portfolio(int id)
      : this(id, 0)
    {
    }

    public Portfolio(int id, int size)
      : this(id, size, null)
    {
    }

    public Portfolio(int id, int size, string[] names)
    {
      m_names = names;
      m_id = id;
      m_pkid = id.ToString();
      m_status = (id % 2 == 0) ? "active" : "inactive";
      m_type = "type" + (id % 3);
      int numSecIds = m_secIds.Length;
      m_position1 = new Position(m_secIds[Position.Count % numSecIds],
        Position.Count * 1000);
      if (id % 2 != 0)
      {
        m_position2 = new Position(m_secIds[Position.Count % numSecIds],
          Position.Count * 1000);
      }
      else
      {
        m_position2 = null;
      }
      m_positions = new Dictionary<Object, Object>();
      m_positions[(m_secIds[Position.Count % numSecIds])] = m_position1;
      if (size > 0)
      {
        m_newVal = new byte[size];
        for (int index = 0; index < size; index++)
        {
          m_newVal[index] = (byte)'B';
        }
      }
      m_creationDate = DateTime.Now;
      m_arrayNull = null;
      m_arrayZeroSize = new byte[0];
    }

    #endregion

    #region IGFSerializable Members

    public IGFSerializable FromData(DataInput input)
    {
      m_id = input.ReadInt32();
      m_pkid = input.ReadUTF();
      m_position1 = (Position)input.ReadObject();
      m_position2 = (Position)input.ReadObject();
      m_positions = new Dictionary<object,object>();
      input.ReadDictionary((System.Collections.IDictionary)m_positions);
      m_type = input.ReadUTF();
      m_status = input.ReadUTF();
      m_names = (string[])(object)input.ReadObject();
      m_newVal = input.ReadBytes();
      //m_creationDate = (System.DateTime)input.ReadGenericObject();
      m_creationDate = input.ReadDate();
      m_arrayNull = input.ReadBytes();
      m_arrayZeroSize = input.ReadBytes();
      
      return this;
    }

    public void ToData(DataOutput output)
    {
      output.WriteInt32(m_id);
      output.WriteUTF(m_pkid);
      output.WriteObject(m_position1);
      output.WriteObject(m_position2);
      output.WriteDictionary((System.Collections.IDictionary)m_positions);
      output.WriteUTF(m_type);
      output.WriteUTF(m_status);
      output.WriteObject(m_names);
      output.WriteBytes(m_newVal);
      //output.WriteObject((IGFSerializable)(object)m_creationDate); // VJR: TODO
      //output.WriteObject(CacheableDate.Create(m_creationDate));
      output.WriteDate(m_creationDate);
      output.WriteBytes(m_arrayNull);
      output.WriteBytes(m_arrayZeroSize);
    }
    
  public UInt32 ObjectSize
    {
      get
      {
        UInt32 objectSize = 0;
        objectSize += (UInt32)sizeof(Int32);
        objectSize += (UInt32)(m_pkid.Length * sizeof(char));
        objectSize += GetObjectSize(m_position1);
        objectSize += GetObjectSize(m_position2);
        objectSize += (UInt32)(m_type.Length * sizeof(char));
        objectSize += (UInt32)(m_status == null ? 0 : sizeof(char) * m_status.Length);
        objectSize += (uint)m_names.Length;//TODO:need to calculate properly
        objectSize += (UInt32)(m_newVal == null ? 0 : sizeof(byte) * m_newVal.Length);
        objectSize += 8; //TODO:need to calculate properly for m_creationDate.;
        objectSize += (UInt32)(m_arrayZeroSize == null ? 0 : sizeof(byte) * m_arrayZeroSize.Length);
        objectSize += (UInt32)(m_arrayNull == null ? 0 : sizeof(byte) * m_arrayNull.Length);
        return objectSize;
      }
    }

    public UInt32 ClassId
    {
      get
      {
        return 0x08;
      }
    }

    #endregion

    public static IGFSerializable CreateDeserializable()
    {
      return new Portfolio();
    }
  }
}
